notify-debouncer-mini/src/lib.rs

```diff
@@ -253,87 +289,91 @@ impl<T: Watcher> Debouncer<T> {
-                    "Invalid tick_rate, tick rate {:?} > {:?} timeout!",
-                    v, timeout
-                ))));
-            }
-            v
-        }
-        None => timeout.checked_div(tick_div).ok_or_else(|| {
-            Error::new(ErrorKind::Generic(format!(
-                "Failed to calculate tick as {:?}/{}!",
-                timeout, tick_div
-            )))
-        })?,
-    };
+    let (tx, rx) = std::sync::mpsc::channel();
 
-    {
-        let mut data_w = data.lock().unwrap();
-        data_w.timeout = timeout;
-    }
-
-    let data_c = data.clone();
-    let stop_c = stop.clone();
     let thread = std::thread::Builder::new()
         .name("notify-rs debouncer loop".to_string())
-        .spawn(move || loop {
-            if stop_c.load(Ordering::Acquire) {
-                break;
-            }
-            std::thread::sleep(tick);
-            let send_data;
-            let errors: Vec<crate::Error>;
-            {
-                let mut lock = data_c.lock().expect("Can't lock debouncer data!");
-                send_data = lock.debounced_events();
-                errors = lock.errors();
-            }
-            if send_data.len() > 0 {
-                event_handler.handle_event(Ok(send_data));
-            }
-            if errors.len() > 0 {
-                event_handler.handle_event(Err(errors));
+        .spawn(move || {
+            let mut data = DebounceDataInner::new(timeout);
+            let mut run = true;
+            // let time_dbg = Instant::now();
+            while run {
+                let mut debounce = false;
+                match data.next_tick() {
+                    Some(timeout) => {
+                        // println!("{} timeout {:?}",time_dbg.elapsed().as_secs(),timeout);
+                        // wait for wakeup
+                        match rx.recv_timeout(timeout) {
+                            Ok(InnerEvent::NotifyEvent(e)) => match e {
+                                Ok(v) => data.add_event(v),
+                                Err(e) => event_handler.handle_event(Err(e)),
+                            },
+                            Ok(InnerEvent::Shutdown) => run = false,
+                            Err(RecvTimeoutError::Disconnected) => run = false,
+                            Err(RecvTimeoutError::Timeout) => debounce = true,
+                        }
+                    }
+                    None => match rx.recv() {
+                        Ok(InnerEvent::NotifyEvent(e)) => match e {
+                            Ok(v) => data.add_event(v),
+                            Err(e) => event_handler.handle_event(Err(e)),
+                        },
+                        Ok(InnerEvent::Shutdown) => run = false,
+                        Err(_) => run = false,
+                    },
+                }
+                // only debounce if we actually passed a threshold
+                if !debounce {
```

Comment:
you can dispatch from the `RecvTimeoutError::Timeout` directly, since that's exactly what you're doing anyway.

Reply:
Done, if we don't handle the possibility of rapid fired events, it makes sense.